### PR TITLE
Switch to constant-time comparison recovery token

### PR DESCRIPTION
**Pull request by cipherboy**
## Description

Recovery mode is a privileged mode whereby operators can directly perform underlying storage operations on the instance. This requires use of recovery key shares to build a singular recovery token which protects API operations. As reported in Vault v2.0.3 changelog (**without corresponding CVE assignment**), the comparison of the recovery token was not constant-time.

We address that and expand unit tests to ensure adequate coverage of edge conditions in token behaviors.

## Rationale

Resolves: GHSA-34fc-gh42-pj53

## Acknowledgements

 - [x] By contributing this change, I certify I have not used generative AI
       (GitHub Copilot, Cursor, Claude Code, &c) in authoring these changes or
       filling out the pull request description or associated issue.
 - [x] By contributing this change, I certify I have signed-off on the
       [DCO ownership](https://developercertificate.org/) statement
       and this change did not use post-BUSL-licensed code from HashiCorp.
       Existing MPL-licensed code is still allowed, subject to attribution.
       Code authored by yourself and submitted to HashiCorp for inclusion is
       also allowed.

### CODE REVIEW

**Comment by cipherboy** [comment]
This requires GHSA assignment; I'm currently leaning towards:


```
CVSS:4.0/AV:N/AC:H/AT:P/PR:N/UI:N/VC:H/VI:H/VA:H/SC:N/SI:N/SA:N
```

= `9.2 CRITICAL`

It requires preconditions (== complexity + requirements) of being in recovery mode, but once there, there's no auditing set up (I believe) and thus can be freely exploited (though, is a timing attack so that's difficult).

File: http/logical.go [hunk 1]
```diff
@@ -328,7 +329,7 @@ func handleLogicalRecovery(raw *vault.RawBackend, token *atomic.Value) http.Hand
 			return
 		}
 		reqToken := r.Header.Get(consts.AuthHeaderName)
-		if reqToken == "" || token.Load() == "" || reqToken != token.Load() {
+		if len(reqToken) == 0 || len(token.Load().(string)) == 0 || subtle.ConstantTimeCompare([]byte(reqToken), []byte(token.Load().(string))) == 0 {
```
Comment (DanGhita) [inline http/logical.go]: I would read the token once and use the variable in the "if"
`current_token: = token.Load()`

**Review by DanGhita — DISMISSED** [review]

File: http/logical.go (hunk 1 above)
Comment (cipherboy) [reply http/logical.go]: Good point @DanGhita, will revise.

**Review by wslabosz-reply — DISMISSED** [review]

**Review by wslabosz-reply — COMMENTED** [review]
fix looks good, although test expansion doesn't really help with the issue; requires a changelog

**Comment by cipherboy** [comment]
@wslabosz-reply said:

> changelog missing :(

Yeah, I'm working on CVE requests for these, will be a minute before we get changelogs (as I want both GHSA and CVE IDs in them).

File: vault/external_tests/misc/recovery_test.go [hunk 2]
```diff
@@ -91,14 +92,31 @@ func TestRecovery(t *testing.T) {
...
+		// Perform an initial request: make sure nothing can be done before a token can be created.
+		client.SetToken("garbage")
+		secret, err := client.Logical().List(path.Join("sys/raw/logical", secretUUID))
+		require.Error(t, err, "expected failure with garbage token before generation")
+		require.Nil(t, secret)
+
+		client.SetToken("")
+		secret, err = client.Logical().List(path.Join("sys/raw/logical", secretUUID))
+		require.Error(t, err, "expected failure with nil token before generation")
+		require.Nil(t, secret)
```
Comment (phil9909) [inline vault/external_tests/misc/recovery_test.go:105]: Too many times I have seen tests which no longer assert what they asserted when they were initially created (and some never did) via asserting "any error should have occurred"

```suggestion
		client.SetToken("garbage")
		secret, err := client.Logical().List(path.Join("sys/raw/logical", secretUUID))
		require.ErrorContains(t, err, "403", "expected failure with garbage token before generation")
		require.Nil(t, secret)

		client.SetToken("")
		secret, err = client.Logical().List(path.Join("sys/raw/logical", secretUUID))
		require.ErrorContains(t, err, "403", "expected failure with nil token before generation")
		require.Nil(t, secret)
```

File: vault/external_tests/misc/recovery_test.go [hunk 3]
```diff
@@ -91,14 +92,31 @@ func TestRecovery(t *testing.T) {
...
+
 		recoveryToken := testhelpers.GenerateRoot(t, cluster, testhelpers.GenerateRecovery)
 		_, err = testhelpers.GenerateRootWithError(t, cluster, testhelpers.GenerateRecovery)
 		if err == nil {
 			t.Fatal("expected second generate-root to fail")
 		}
```
Comment (phil9909) [inline vault/external_tests/misc/recovery_test.go]:
```suggestion
		require.ErrorContains(t, err, "attempted to generate recovery operation token when already unsealed", "expected second generate-root to fail")
```

File: vault/external_tests/misc/recovery_test.go [hunk 4]
```diff
@@ -91,14 +92,31 @@ func TestRecovery(t *testing.T) {
...
+
+		secret, err = client.Logical().List(path.Join("sys/raw/logical", secretUUID))
+		require.Error(t, err, "expected failure with nil token after generation")
```
Comment (phil9909) [inline vault/external_tests/misc/recovery_test.go]:
```suggestion
		require.ErrorContains(t, err, "403", "expected failure with nil token after generation")
```

sorry for splitting this into multiple comments, but GitHub gets confused over suggestions which span multiple hunks.

**Review by phil9909 — DISMISSED** [review]

**Review by phil9909 — DISMISSED** [review]

**Comment by cipherboy** [comment]
> Yeah, I'm working on CVE requests for these, will be a minute before we get changelogs (as I want both GHSA and CVE IDs in them).

So much for this. Been over two weeks and @GitHub hasn't assigned CVE identifiers :laughing:

**Comment by cipherboy** [comment]
DCO seems to not understand where `main` is at? :thinking: 

image

**Review by wslabosz-reply — APPROVED** [review]
LGTM

**Comment by cipherboy** [comment]
Thanks @wslabosz-reply, merging!
